### DISCUSSION TITLE
Product categories

### DISCUSSION THREAD

**Discussion by de-raaf-media** [duncanmcclean/simple-commerce#914] [Help] [answered]
Hi,

I am trying out simple commerce on my development environment. I have a question. In GitHub issues I see product categories are mentioned. But I do not see an option to add a category to a product via the CP. What is the correct way to an add a product to a category? Or is this not part of simple-commerce and should I change the blueprint for instance?

Regards,
Mark

**Comment by duncanmcclean** [comment, marked as answer]
Hello!

Product Categories only existed very early on in Simple Commerce's existence, they've since been removed. 

As products are just normal entries, you can create a taxonomy for product categories then link them up that way.

**Reply by de-raaf-media** [reply]
Thanks @duncanmcclean, makes sense, 

I actually need this so I can filter products. Before building this from scratch: is there a plugin (or example implementation) of a product filter for a front-end? So for instance clicking on blue will only show blue jeans? I think almost every webshop needs this so I can imagine this exists already.

**Reply by duncanmcclean** [reply]
I'm not sure if there is anything existing for that but you could build it with something like [Livewire](https://statamic.com/addons/jonassiewertsen/livewire) or just plain Antlers and query parameters.

This might be more of a discussion to have with folks on the Statamic Discord, I'm sure others there will have done similar things.

**Reply by de-raaf-media** [reply]
Allright. Good to know. I will look into it. Thanks!

**Reply by de-raaf-media** [reply]
Hi @duncanmcclean, I am trying to filter products by product_variants options. In eloquent I know how to achieve this (maybe I will convert my collection to database tables). But I cannot seem to find the right way of doing this with the Statamic entry. Can you point me in the right direction maybe?

$entries = Entry::query()
            ->where('collection', 'products')
            ->get();

**Reply by duncanmcclean** [reply]
At the moment, you can do something like this...

```php
$entries = Entry::query()
    ->where('collection', 'products')
    ->where('product_variants->options->0->key', 'Blue_Small')
    ->orWhere('product_variants->options->1->key', 'Blue_Small')
    ->get()
```

In this example, it's queuing products where the first or second variant option is `Blue_Small`. 

Right now, there's no way to do wildcards with nested fields like that, so you'll have to do a separate `->orWhere` line for third/fourth/fifth/etc options on a product. Wildcards may be coming (it's a core Statamic thing), see https://github.com/statamic/cms/pull/7454 and https://github.com/statamic/cms/pull/8476.

**Reply by de-raaf-media** [reply]
Hm ... you would have to know the number of variant options in that case. Wildcards would improve this a lot.

I think for now I will go for the approach of migrating to a database and perform the query for the product filtering there.

Thanks Duncan

**Reply by duncanmcclean** [reply]
> I think for now I will go for the approach of migrating to a database and perform the query for the product filtering there.

Are you planning on moving products to your own Eloquent models or using Statamic's Eloquent Driver? 

If you go down the route of using your own Eloquent model, you'll need to implement your own [ProductRepository](https://github.com/duncanmcclean/simple-commerce/blob/5.x/src/Products/EntryProductRepository.php) to handle finding/updating products in the database, rather than with entries. Simple Commerce doesn't include products as part of it's "database" setup due to the increased maintenance overhead.

**Comment by de-raaf-media** [comment]
Actually - I may have been going in the wrong direction. Statamic is still a bit new to me. You mentioned taxonomies before. By using taxonomies - and combining them - I now implemented a collection filter with a couple of categories. So instead of filtering on Product Variants I am going to filter on one or more taxonomies.

taxonomy:categories="{get:categories}"
 
I combined your sort mechanism with product categories and I think this is turning out to be quite a flexible solution that I can keep extending with different taxonomies.

To answer your question: It was my intention to use the Statamic Eloquent Driver. And then use a separate model for querying and returning json to some kind of javascript filter implementation. But I am not going for that solution at this time.

Thanks for you help and input. Very helpful.

**Reply by duncanmcclean** [reply]
No worries - glad you've got it sorted!
